lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -3395,16 +3395,82 @@ void RegResetOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
-    return (*this)->emitOpError("invalid symbol reference");
+FormalOp::verifySymbolUses(mlir::SymbolTableCollection &symbolTable) {
+  auto *op = symbolTable.lookupNearestSymbolFrom(*this, getModuleNameAttr());
+  if (!op)
+    return emitOpError() << "targets unknown module " << getModuleNameAttr();
+
+  if (!isa<FModuleLike>(op)) {
+    auto d = emitOpError() << "target " << getModuleNameAttr()
+                           << " is not a module";
+    d.attachNote(op->getLoc()) << "target defined here";
+    return d;
+  }
 
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// SimulationOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult
+SimulationOp::verifySymbolUses(mlir::SymbolTableCollection &symbolTable) {
+  auto *op = symbolTable.lookupNearestSymbolFrom(*this, getModuleNameAttr());
+  if (!op)
+    return emitOpError() << "targets unknown module " << getModuleNameAttr();
+
+  auto complain = [&] {
+    auto d = emitOpError() << "target " << getModuleNameAttr() << " ";
+    d.attachNote(op->getLoc()) << "target defined here";
+    return d;
+  };
+
+  auto module = dyn_cast<FModuleLike>(op);
+  if (!module)
+    return complain() << "is not a module";
+
+  auto numPorts = module.getPortDirections().size();
+  if (numPorts != 4)
+    return complain() << "must have 4 ports, got " << numPorts << " instead";
+
+  auto checkPort = [&](unsigned idx, StringRef expName, Direction expDir,
+                       llvm::function_ref<bool(Type)> checkType,
+                       StringRef expType) {
+    auto name = module.getPortNameAttr(idx);
+    if (name != expName) {
+      complain() << "port " << idx << " must be called \"" << expName
+                 << "\", got " << name << " instead";
+      return false;
+    }
+    if (auto dir = module.getPortDirection(idx); dir != expDir) {
+      auto stringify = [](Direction dir) {
+        return dir == Direction::In ? "an input" : "an output";
+      };
+      complain() << "port " << name << " must be " << stringify(expDir)
+                 << ", got " << stringify(dir) << " instead";
+      return false;
+    }
+    if (auto type = module.getPortType(idx); !checkType(type)) {
+      complain() << "port " << name << " must be a '!firrtl." << expType
+                 << "', got " << type << " instead";
+      return false;
+    }
+    return true;
+  };
+
+  auto isClock = [](Type type) { return isa<ClockType>(type); };
+  auto isBool = [](Type type) {
+    if (auto uintType = dyn_cast<UIntType>(type))
+      return uintType.getWidth() == 1;
+    return false;
+  };
+  return success(checkPort(0, "clock", Direction::In, isClock, "clock") &&
+                 checkPort(1, "init", Direction::In, isBool, "uint<1>") &&
+                 checkPort(2, "done", Direction::Out, isBool, "uint<1>") &&
+                 checkPort(3, "success", Direction::Out, isBool, "uint<1>"));
+}
```

Comment:
This verification is annoyingly duplicated for all simulation ops.

There are two ways to resolve this:

1. The verification is inverted. I.e., this port restriction is moved onto the module. The module then looks for symbol users and if it is used by a simulation op, then it checks its port structure. I'm not sure if this kind of verification is allowed.
2. The IR for this should be changed so that the simulation op is a list of tests (or something else which collapses this down).

Reply:
I think (1) might have worse performance than the existing solution, since the `verifSymbolUses` gets a SymbolTable to make the symbol lookup efficient. If we flipped that around, the verifier for `FModuleOp` would have to scan the entire IR for uses (can't use cached symbol user maps in a verifier). It feels wrong to have verification for `firrtl.simulation` split and put onto `firrtl.module`.

(2) could work! I'm pretty sure the performance overhead here is negligible, though. The input will have roughly 1 simulation op per module, maybe a handful in the worst case. I think that pales in comparison to the port list verification that frequent `func.call` and `hw.instance` ops would have to do.